pkg/image/apis/image/validation/whitelist/whitelister.go

```diff
@@ -0,0 +1,268 @@
+	// the given glob will match hosts with both :80 and :443 ports.
+	WhitelistRegistry(hostPortGlob string, transport WhitelistTransport) error
+	// WhitelistPullSpecs allows to whitelist particular pull specs. References must match exactly one of the
+	// given pull specs for it to be whitelisted.
+	// TODO: accept insecure flag
+	WhitelistPullSpecs(pullSpecs ...string)
+	// Copy returns a deep copy of the whitelister. This is useful for temporarily whitelisting additional
+	// registries/pullSpecs before a specific validation.
+	Copy() RegistryWhitelister
+}
+
+type allowedHostPortGlobs struct {
+	host string
+	port string
+}
+
+type registryWhitelister struct {
+	whitelist             []allowedHostPortGlobs
+	pullSpecs             sets.String
+	registryHostRetriever imageapi.RegistryHostnameRetriever
+}
+
+var _ RegistryWhitelister = &registryWhitelister{}
+
+// NewRegistryWhitelister creates a whitelister that admits registry domains and pull specs based on the given
+// list of allowed registries and the current domain name of the integrated Docker registry.
+func NewRegistryWhitelister(
+	whitelist serverapi.AllowedRegistries,
+	registryHostRetriever imageapi.RegistryHostnameRetriever,
+) (RegistryWhitelister, error) {
+	errs := []error{}
+	rw := registryWhitelister{
+		whitelist:             make([]allowedHostPortGlobs, 0, len(whitelist)),
+		pullSpecs:             sets.NewString(),
+		registryHostRetriever: registryHostRetriever,
+	}
+	// iterate in reversed order to make the patterns appear in the same order as given (patterns are prepended)
+	for i := len(whitelist) - 1; i >= 0; i-- {
+		registry := whitelist[i]
+		err := rw.WhitelistRegistry(registry.DomainName, GetWhitelistTransportForFlag(registry.Insecure, false))
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		return nil, kerrutil.NewAggregate(errs)
+	}
+	return &rw, nil
+}
+
+// WhitelistAllRegistries returns a whitelister that will allow any given registry host name.
+// TODO: make a new implementation of RegistryWhitelister instead that will not bother with pull specs
+func WhitelistAllRegistries() RegistryWhitelister {
+	return &registryWhitelister{
+		whitelist: []allowedHostPortGlobs{{host: "*", port: "*"}},
+		pullSpecs: sets.NewString(),
+	}
+}
+
+func (rw *registryWhitelister) AdmitHostname(hostname string, transport WhitelistTransport) error {
+	return rw.AdmitDockerImageReference(&imageapi.DockerImageReference{Registry: hostname}, transport)
+}
+
+func (rw *registryWhitelister) AdmitPullSpec(pullSpec string, transport WhitelistTransport) error {
+	ref, err := imageapi.ParseDockerImageReference(pullSpec)
+	if err != nil {
+		return err
+	}
+	return rw.AdmitDockerImageReference(&ref, transport)
+}
+
+func (rw *registryWhitelister) AdmitDockerImageReference(ref *imageapi.DockerImageReference, transport WhitelistTransport) error {
+	const showMax = 5
+	const fullWhitelistLogLevel = 5
```

Comment:
This is really uncommon, I would prefer you not do this because it is inconsistent with the rest of the code base.  It turns a readable log line into an unreadable log line.

Reply:
I was worried about polluting the logs when there is 15+ whitelisted registries. I can certainly remove it.

Reply:
I think @smarterclayton just wants you to use "5" in place and not define a constant for it, he's not asking you to actually change the level you're logging at. I don't think anyway.